File: r/generate_results.R
# PURPOSE ----
# Generate normalized R results for every shared conformance fixture.

library(sbgnvalidator)

# FUNCTIONS ----

#' Write one normalized validation report.
#'
#' @param schema_cache Environment containing compiled schemas.
#' @param benchmark_root Absolute benchmark directory.
#' @param repository_root Absolute repository directory.
#' @param entry One fixture manifest entry.
write_result <- function(schema_cache, benchmark_root, repository_root, entry) {
  schema_path <- file.path(repository_root, entry$schema)
  document_path <- file.path(repository_root, entry$input)
  if (identical(entry$namespace_policy, "allow-sbgnml-0.2")) {
    report <- validate_sbgn(
      document_path,
      rules_path = schema_path,
      phase = entry$phase,
      allow_sbgnml_0_2 = TRUE
    )
  } else {
    cache_key <- paste(schema_path, entry$phase, sep = "::")
    if (!exists(cache_key, envir = schema_cache, inherits = FALSE)) {
      assign(
        cache_key,
        schematron_compile(schema_path, entry$phase),
        envir = schema_cache
      )
    }
    report <- schematron_validate(
      get(cache_key, envir = schema_cache, inherits = FALSE),
      document_path
    )
  }
  output_path <- file.path(
    benchmark_root,
    "build",
    "results",
    "r",
    sub("^conformance/oracle/java/", "", entry$oracle)
  )
  dir.create(dirname(output_path), recursive = TRUE, showWarnings = FALSE)
  jsonlite::write_json(
    report,
    output_path,
    auto_unbox = TRUE,
    pretty = TRUE,
    null = "null"
  )
}

# LOAD DATA ----

benchmark_root <- normalizePath("..", mustWork = TRUE)
repository_root <- benchmark_root
manifest <- jsonlite::read_json(
  file.path(benchmark_root, "conformance", "manifest.json"),
  simplifyVector = FALSE
)

# ANALYSIS ----

schema_cache <- new.env(parent = emptyenv())
for (entry in manifest$cases) {
  write_result(schema_cache, benchmark_root, repository_root, entry)
}
message("generated ", length(manifest$cases), " R reports")

File: rust/src/bin/generate_results.rs
use sbgn_validator::{NamespacePolicy, Validator};
use serde::Deserialize;
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Deserialize)]
struct Manifest {
    cases: Vec<ManifestCase>,
}

#[derive(Deserialize)]
struct ManifestCase {
    schema: String,
    phase: String,
    input: String,
    oracle: String,
    namespace_policy: Option<String>,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let benchmark_root = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("..");
    let repository_root = benchmark_root.clone();
    let manifest: Manifest = serde_json::from_str(&fs::read_to_string(
        benchmark_root.join("conformance/manifest.json"),
    )?)?;
    let mut validators = BTreeMap::new();
    for case in &manifest.cases {
        let document = repository_root.join(&case.input);
        let report = if case.namespace_policy.as_deref() == Some("allow-sbgnml-0.2") {
            Validator::compile_for_document(
                &repository_root.join(&case.schema),
                &document,
                &case.phase,
                NamespacePolicy::AllowSbgnml02,
            )?
            .validate(&document)?
        } else {
            let key = format!("{}#{}", case.schema, case.phase);
            if !validators.contains_key(&key) {
                validators.insert(
                    key.clone(),
                    Validator::compile(&repository_root.join(&case.schema), &case.phase)?,
                );
            }
            validators
                .get(&key)
                .expect("validator was just inserted")
                .validate(&document)?
        };
        let relative = case
            .oracle
            .strip_prefix("conformance/oracle/java/")
            .ok_or("oracle path is outside the Java oracle directory")?;
        let output = benchmark_root.join("build/results/rust").join(relative);
        fs::create_dir_all(output.parent().unwrap_or(Path::new(".")))?;
        fs::write(output, serde_json::to_string_pretty(&report)? + "\n")?;
    }
    println!("generated {} Rust reports", manifest.cases.len());
    Ok(())
}
